### PR TITLE
build: Add additional step nodes when labels are modified

### PR DESCRIPTION
**Pull request by flouthoc**
Make sure label modification behaviour stays similar to docker i.e
new step is added when labels are modified and leave caching of layers
to executor

Advantages.

* Layers will be automatically cached if needed.
* If Labels are modified , new step will be built and cached if
  configured instead of forcing pre-built cached layers.
  
See detailed discussion here: https://github.com/containers/buildah/issues/3370#issuecomment-918024546

Closes: https://github.com/containers/buildah/issues/3370

### CODE REVIEW

**Comment by flouthoc** [comment]
Will add tests once we finalize on this discussion.

**Comment by flouthoc** [comment]
@containers/buildah-maintainers @nalind PTAL

**Comment by nalind** [comment]
Looks reasonable.  Tests should definitely make sure this doesn't break weirdly on multistage builds and when the build replaces the base image (i.e., when the `--from` CLI option is used).  Please take care around quoting label values.

File: imagebuildah/build.go [hunk 1]
```diff
@@ -318,6 +318,37 @@ func buildDockerfilesOnce(ctx context.Context, store storage.Store, logger *logr
...
+			labelLine := ""
+			key := label[0]
+			value := ""
+			if len(label) > 1 {
+				value = label[1]
+			} else {
+				value = ""
+			}
+			labelLine = fmt.Sprintf("LABEL %s=%s\n", key, value)
+			if len(labelLine) > 1 {
```
Comment (nalind) [inline imagebuildah/build.go:341]: When would this be `false`?

File: imagebuildah/build.go [hunk 2]
```diff
@@ -318,6 +318,37 @@ func buildDockerfilesOnce(ctx context.Context, store storage.Store, logger *logr
...
+				labelLineBytes := []byte(labelLine)
+				additionalNode, err := imagebuilder.ParseDockerfile(bytes.NewReader(labelLineBytes))
```
Comment (nalind) [inline imagebuildah/build.go:343]: Woudl `strings.NewReader()` work better than casting to a byte slice and using `bytes.NewReader()` here?

File: imagebuildah/build.go [hunk 3]
```diff
@@ -318,6 +318,37 @@ func buildDockerfilesOnce(ctx context.Context, store storage.Store, logger *logr
...
+		for _, labelSpec := range options.Labels {
+			label := strings.SplitN(labelSpec, "=", 2)
+			labelLine := ""
+			key := label[0]
+			value := ""
+			if len(label) > 1 {
+				value = label[1]
+			} else {
+				value = ""
+			}
```
Comment (TomSweeneyRedHat) [inline imagebuildah/build.go:339]: I don't think the else is necessary as you're setting the default at line 334

File: imagebuildah/build.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply imagebuildah/build.go:341]: yeah, might be better to check if key and value don't equal ""

**Comment by rhatdan** [comment]
@flouthoc are you planning on updating this PR?

**Comment by flouthoc** [comment]
@rhatdan Just writing some test cases which covers all edge-cases.

File: imagebuildah/build.go (hunk 1 above)
Comment (flouthoc) [reply imagebuildah/build.go:339]: I think this is needed for basic validation since `--label` accepts arbitrary string similar check is also present in executor itself.

**Comment by flouthoc** [comment]
@nalind  added tests for `regular` and `multi-staged` builds

**Muti-stage build test covers**
* Perform multi-stage build with `label1=value1` and verify labels.
* Relabel build with `label1=value2` and verify
* To make sure we are using cache rebuild with `label1=value1`  and makes sure everything is used from cache

**Regular build with --from**
* Covers basic modifying of labels and base images

Just waiting for everything to pass.

File: imagebuildah/build.go [hunk 4]
```diff
@@ -318,6 +318,34 @@ func buildDockerfilesOnce(ctx context.Context, store storage.Store, logger *logr
...
+	if len(options.Labels) > 0 {
+		for _, labelSpec := range options.Labels {
+			label := strings.SplitN(labelSpec, "=", 2)
+			labelLine := ""
+			key := label[0]
+			value := ""
+			if len(label) > 1 {
+				value = label[1]
+			}
+			labelLine = fmt.Sprintf("LABEL %s=%s\n", key, value)
```
Comment (nalind) [inline imagebuildah/build.go:338]: Really do need to look at quoting here.

Comment (flouthoc) [reply imagebuildah/build.go:338]: @nalind ah yes, adding this 
```diff
-                       labelLine = fmt.Sprintf("LABEL %s=%s\n", key, value)
+                       labelLine = fmt.Sprintf("LABEL \"%s\"=\"%s\"\n", key, value)

File: imagebuildah/build.go [hunk 5]
```diff
@@ -318,6 +318,34 @@ func buildDockerfilesOnce(ctx context.Context, store storage.Store, logger *logr
...
+	if len(options.Labels) > 0 {
+		for _, labelSpec := range options.Labels {
+			label := strings.SplitN(labelSpec, "=", 2)
+			labelLine := ""
+			key := label[0]
+			value := ""
+			if len(label) > 1 {
+				value = label[1]
+			}
+			labelLine = fmt.Sprintf("LABEL \"%s\"=\"%s\"\n", key, value)
```
Comment (nalind) [inline imagebuildah/build.go:338]: This doesn't look like it'll catch weirdness like values that include internal quotation marks of one kind or another.

Comment (flouthoc) [reply imagebuildah/build.go:338]: @nalind i tried this `buildah build --layers --label "label1=value'2" --from=alpine -t image1`  but this is failing  `buildah build --layers --label 'label1=value"2' --from=alpine -t image1` let me get a fix.

Comment (flouthoc) [reply imagebuildah/build.go]: @nalind I have changed this formatting with simple `%q` it autoescapes things for us tested both the cases, `"` will be autoescaped. 
```diff
-                       labelLine = fmt.Sprintf("LABEL \"%s\"=\"%s\"\n", key, value)
+                       labelLine = fmt.Sprintf("LABEL %q=%q\n", key, value)
```

**Comment by rhatdan** [comment]
/approve
LGTM

File: imagebuildah/build.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply imagebuildah/build.go:339]: Right, but isn't labelLine at a minimum going to contain "LABEL = "?  That would never be 0 or less right?

**Comment by flouthoc** [comment]
Just a rebase since main is updated.

File: imagebuildah/build.go (hunk 1 above)
Comment (flouthoc) [reply imagebuildah/build.go]: @TomSweeneyRedHat oh right thanks changing this to `key != ""`, we can support empty value since docker supports empty values (manually verified).

**Comment by TomSweeneyRedHat** [comment]
LGTM
I'm not sure if this needs a rebase or not....
